hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/functional/TestHoodieBackedMetadata.java

```diff
@@ -480,6 +482,131 @@ public void testRollbackUnsyncedCommit(HoodieTableType tableType) throws Excepti
       client.syncTableMetadata();
       validateMetadata(client);
     }
+
+    // If an unsynced commit is automatically rolled back during next commit, the rollback commit gets a timestamp
+    // greater than than the new commit which is started. Ensure that in this case the rollback is not processed
+    // as the earlier failed commit would not have been committed.
+    //
+    //  Dataset:   C1        C2         C3.inflight[failed]   C4   R5[rolls back C3]
+    //  Metadata:  C1.delta  C2.delta
+    //
+    // When R5 completes, C3.xxx will be deleted. When C4 completes, C4 and R5 will be committed to Metadata Table in
```

Comment:
I think we are doing this just for sake of maintaining this invariant that, deleting a file that does not exist should be a fatal exception. I think we can just relax this, instead of introducing more complexity.

Reply:
When we relax this we cannot catch issues wherein we may have missed syncing an instant earlier. See HUDI-2286.

Reply:
On this one, I continue to disagree :). We can easily do a logger.warn or error and collect information in an automated fashion, to fix the issues. In the current model, heavy user intervention is needed to fix the metadata table and then get the pipeline back online. This is not very user-friendly for folks in OSS.

I am thinking of adding an internal config to make this behavior configurable, so you can have it failing hard in Uber and we can do the other way in OSS. It ll actually help us see more variety of issues and harden the implementation. wdyt?

Reply:
Having a config is acceptable as there can be various use cases for HUDI.

We saw a production issue where the metadata table did not have the latest files. If there was no exception:
1. Only an error would have been logged
2. Metadata table would have returned an older file listing
3. Data would have got written to older version of files

The above is a classic data lass scenario which would be very difficult to catch once the logs have rolled over.  Hence, I want to err on the side of data-consistency.

To contrast it to another component in HUDI wherein we do something similar - If the RemoteFileSystemView is unable to get the listing from TimelineServer, it logs an error and then falls back to file listing. There are no data consistency issues in this fallback so this makes sense to prevent user interventions. But wrong results from metadata table can lead to data consistency issues.

BTW, there may be a better way to sync the rollback. Also much of this complexity will go away with the synchronous design.

Reply:
okay. lets revisit after the sync design and I ll add a flag down the line if needed.